url8e6440c8aa_robdimsdale_project_euler_tgz-pJ8-com_rmd_personal_projecteuler_MainJ8/src/com/rmd/personal/projecteuler/p1to19/P005.java

```diff
@@ -0,0 +1,56 @@
+package com.rmd.personal.projecteuler.p1to19;
+
+import com.rmd.personal.projecteuler.Common;
+import com.rmd.personal.projecteuler.Problem;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class P005 implements Problem {
+
+    private static final int DEFAULT_MAX_FACTOR = 20;
+    private int maxFactor;
+
+    public P005() {
+        this.setMaxFactor(DEFAULT_MAX_FACTOR);
+    }
+
+    public int getMaxFactor() {
+        return maxFactor;
+    }
+
+    public void setMaxFactor(int maxFactor) {
+        this.maxFactor = maxFactor;
+    }
+
+    @Override
+    public String getDescription() {
+        return "2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any "
+                + "remainder.\n\n"
+                + "What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?";
+    }
+
+    @Override
+    public String run() {
+        Map<Long, Integer> primePowers = new HashMap<Long, Integer>();
+
+        long currentPrime = 2;
+        while (currentPrime < this.getMaxFactor()) {
+            int currentPrimePower = 0;
+            while (Math.pow(currentPrime, currentPrimePower + 1) < this.getMaxFactor()) {
+                currentPrimePower++;
+            }
+            if (primePowers.get(currentPrime) == null || primePowers.get(currentPrime) < currentPrimePower) {
+                primePowers.put(currentPrime, currentPrimePower);
+            }
+            currentPrime = Common.getPrimes().get(Common.getPrimes().indexOf(currentPrime) + 1);
+        }
+
+        double product = 1;
+        for (long prime : primePowers.keySet()) {
+            product = product * Math.pow(prime, primePowers.get(prime));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.